### ISSUE TITLE
add a type definition to the package

### ISSUE THREAD

**Issue by MHA15** [zloirock/core-js#568]
hello

I know that there is already a typescript type definition for this package in `@types/core-js` but many packages are trying to make a type definition inside of they're packages which has benefits like always sync to the package version and doesn't need to install it separately 

can you add a type definition into this repo ? and I'm interested to do it if you want

**Comment by nicolo-ribaudo** [comment]
Since core-js matches the JavaScript API, doesn't TypeScript support them out of the box?

**Comment by MHA15** [comment]
it doesn't match completely , it have some proposal functions that not in the typescript lib, and also some interfaces like Dict which are not in pure js

**Comment by zloirock** [comment]
@nicolo-ribaudo you're right, but only about stable JS and global polyfills. `core-js` contains also polyfills for early-stage proposals which haven't any standard type definitions and entry points for a version without global pollution. Type definitions for them could be useful.

@MHA15 `Dict` and other non-standard features removed from `core-js@3`.

Feel free to add a PR.

**Comment by zloirock** [comment]
https://github.com/Kingwl/core-dts

**Comment by zloirock** [comment]
Type definitions for `core-js-builder` added in `3.28.0`.

**Comment by lionel-rowe** [comment]
There's probably quite a bit of work to get this all set up, but it'd be great if behavior was as follows:

* Imported classes/functions from any `core-js-pure/*` entrypoint are typed appropriately, with no ambient types
* Import from any `core-js/*` entrypoint ambiently loads the types, but _only_ the types that are relevant to that entrypoint, e.g. `core-js-pure/full/string/dedent.js` ambiently loads `dedent` method under `StringConstructor` interface (but nothing else)

It'd also be good to have some level of auto-generated or semi-auto-generated "minimal" support as a fallback for modules without proper type definitions yet.

Current:

```ts
import mod from 'core-js-pure/full/index.js'
// type Mod = any
type Mod = typeof zip
```

Better ("minimal" support):

```ts
import mod from 'core-js-pure/full/index.js'
// type Mod = {
//     ...
//     String: {
//         ...
//         dedent(...args: any[]): any,
//         ...
//     },
//     ...
// }
type Mod = typeof zip
```

Best (full support):

```ts
import mod from 'core-js-pure/full/index.js'
// type Mod = {
//     ...
//     String: {
//         ...
//         dedent(str: string): string;
//         dedent(str: TemplateStringsArray, ...substitutions: unknown[]): string;
//         dedent<A extends unknown[], R, T>(tag: Tag<A, R, T>): Tag<A, R, T>;
//         ...
//     },
//     ...
// }
type Mod = typeof zip
```

**Comment by slowcheetah** [comment]
@lionel-rowe you’re describing something really close to how our current implementation works. Right now, we have global type extensions, and for each entry point we build an exported type from those globals, which then gets automatically included when you import that entry point. 

For example, type for [Iterator chunking proposal](https://github.com/zloirock/core-js/blob/v4-types/packages/core-js/types/proposals/iterator-chunking.d.ts). When the bundle is built, a .d.ts file will be created next to each entry point, and it’ll look like:

```
import '../../types/proposals/iterator-chunking';

type method = Iterator<any>['chunks'];
type uncurriedMethod<T> = (self: Iterator<T>, ...args: Parameters<method>) => ReturnType<method>;
declare const resultMethod: uncurriedMethod<any>;

export default resultMethod;
```

You can try it out on the [v4-types branch](https://github.com/zloirock/core-js/tree/v4-types).

Thoughts, ideas or feedback are welcome.

**Comment by slowcheetah** [comment]
When designing type definitions for core-js, we decided to ship them as a separate package, because we cannot guarantee stable behavior, primarily with upcoming minor TypeScript releases. For example, shipping a separate package allows us to build different bundles for different TypeScript versions and automatically provide the appropriate types for the user's TS version.

To avoid polluting the global namespace in the pure version, we chose an approach where the types are wrapped in a namespace and then exposed through modules. For example:
```ts
declare namespace CoreJS {
  interface String {
    padStart(maxLength: number, fillString?: string): string;
  }
}

declare module "@core-js/pure/full/string/pad-start" {
  type method = CoreJs.String['padStart'];
  const resultMethod: (self: CoreJS.String, ...args: Parameters<method>) => ReturnType<method>;
  export = resultMethod;
}
```

During development, we ran into the problem that TypeScript does not support conditional type definitions at the declaration level (i.e. define this type only if it doesn't already exist, otherwise reuse the existing one). This pushed us toward creating "universal" types that do not conflict with the built‑in TS types.

A large part of the package is already written in this style, but there are some dead‑end cases:

- **Conflicts with DOM built‑ins.**
Many DOM types are defined via **declare var**, so TypeScript does not allow extending them.
Example: Suppose we want to add a static parse method to URL. In lib.dom.d.ts, URL is defined as:
```ts
declare var URL: {
  prototype: URL;
  new(url: string | URL, base?: string | URL): URL;
  //  ...
}
```
When we try to redeclare this, TypeScript reports an error that the variable already exists, so we can’t safely augment it.

- **Use of getters and setters.**
There is no way to create a universal type that works cleanly with properties defined through getters/setters. We could model such properties as readonly fields in interfaces, which would be extendable, but in some standard definitions they are written explicitly as getters/setters.
Example: The proposal https://github.com/tc39/proposal-arraybuffer-transfer adds a detached property to ArrayBuffer, but in es2024 it is defined as:
```ts
interface ArrayBuffer {
  // ...
  get detached(): boolean;
  // ...
}
```
Does anyone have ideas on how these limitations could be worked around, and what do you think about the overall approach we've chosen?

**Comment by DetachHead** [comment]
i can't seem to reproduce either of those issues, but if you're encountering conflicts with the builtin DOM types, perhaps the [`libReplacement`](https://www.typescriptlang.org/tsconfig/#libReplacement) setting in `tsconfig.json` might help?

**Comment by zloirock** [comment]
`v4.0.0-alpha.1` of [`@core-js/types`](https://www.npmjs.com/package/@core-js/types) is published.
